### ISSUE TITLE
fill-paragraph breaks code structure

### ISSUE THREAD

**Issue by andreyorst** [justinbarclay/parinfer-rust-mode#9]
Sorry if I'm overwhelming you with issues, but I've finally got my hands on your package, and had put it to the test.

Here's the recording https://asciinema.org/a/ykl6Brm5oH8K49s5R45Elsw25

What I'm doing is this: when cursor is outside the doc string I'm pressing `C-q` that runs `fill-paragraph` function, which hard wraps text. This works correct for both cases when cursor is before doc string and after doc string. But when it is inside doc string the structure of above code changes.

This doesn't seem to happen in non Clojure buffers. Here's example code:

```clojure
(ns some-namespace.core
  (:import [java.util.concurrent Executors])
  (:require [clojure.string :as string]))

(defn foo
  "pretty long docsting that will be affected by `fill-paragraph` function"
  [x]
  x)
```

**Comment by justinbarclay** [comment]
No need to apologize, these are all good issues. These videos are very insightful too, smart-mode is a complex beast so it helps to _see_ where it's breaking.

**Comment by justinbarclay** [comment]
Note for myself when I can look into this: Clojure mode overwrites how fill paragraph works, specifically when it is called inside of a string. It narrows to region on the string block, runs some transformation on text, and then widens to the buffer.

This is troublesome for parinfer-rust-mode because while detecting changes, if a buffer is narrowed, it will see those changes as existing within that narrowed scope and not within the entire file/buffer. So in the example above it will see the changes to the doc string on line 1 instead of line 5. This can be exasperated by it also using the old buffer text, from the full file, and comparing it against the text in the narrowed region/buffer. This will send very confusing data to parinfer-rust.

**Comment by justinbarclay** [comment]
Hey, I've come back to this issue and while running clojure-mode and parinfer-rust-mode (43a49d7), I'm no longer able to reproduce. Are you still running into issues with `fill-paragraph`?

**Comment by andreyorst** [comment]
Still happens to me with the example code I've provided

**Comment by justinbarclay** [comment]
Thank you, that answer helped me figure out why I couldn't reproduce it. I have since been able to reproduce it and come up with a potential fix. This is on another branch that I am going to test over the next couple of days. I'll update this issue when it gets merged in.

**Comment by justinbarclay** [comment]
This issue _should_ be resolved for real this time

**Comment by andreyorst** [comment]
seems to work correct even for narrowed regions. I think it is solved! If I encounter any other place where this happens I'll reopen, but I've tested all places where I've seen this issue, and all of those seem to work fine.

**Comment by andreyorst** [comment]
It happened again. I'll post repro later

**Comment by andreyorst** [comment]
@justinbarclay actually, the initial recipe reproduces the issue

**Comment by justinbarclay** [comment]
I was able to recreate this issue and fixed it with 967dfa92a153f4e82418befe4448e4f328e57455 (which is why I think why my tests  for this case passed but didn't work in a non test environment)

**Comment by andreyorst** [comment]
> I was able to recreate this issue and fixed it with [967dfa9](https://github.com/justinbarclay/parinfer-rust-mode/commit/967dfa92a153f4e82418befe4448e4f328e57455) (which is why I think my tests for this case passed but didn't work in a non test environment)

yes, it works now. Thanks!
